include/mxnet/c_lapack_api.h

```diff
@@ -54,6 +57,40 @@ extern "C" {
 
 inline char loup(char uplo, bool invert) { return invert ? (uplo == 'U' ? 'L' : 'U') : uplo; }
 
+
+/*!
+ * \brief Transpose matrix data in memory
+ * 
+ * Equivalently we can see it as flipping the layout of the matrix
+ * between row-major and column-major.
+ *
+ * \param m number of rows of input matrix a
+ * \param n number of columns of input matrix a
+ * \param b output matrix
+ * \param ldb leading dimension of b
+ * \param a input matrix
+ * \param lda leading dimension of a
+ */
+template <typename xpu, typename DType>
+inline void flip(int m, int n, DType *b, int ldb, DType *a, int lda);
```

Comment:
The `transpose()` defined in `mshadow/extension/transpose.h` operates on templated `Exp` and returns a templated `Exp` so that the compiler can finally instantiate out the entire Expression. What we're doing here is to *actually* transpose the layout of the data storage in memory so that later on LAPACK calls can read/interpret the data correctly.

For LAPACK libs that support only column-major matrices, we'd always need such a helper to transpose the data storage, perform the calculation, and transpose back the result. It seems the easiest way to handle row-major matrices and also the most efficient empirically.